pkg/xds/generator/dns_generator.go

```diff
@@ -0,0 +1,72 @@
+package generator
+
+import (
+	"github.com/asaskevich/govalidator"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	core_xds "github.com/kumahq/kuma/pkg/core/xds"
+	xds_context "github.com/kumahq/kuma/pkg/xds/context"
+	envoy_common "github.com/kumahq/kuma/pkg/xds/envoy"
+	envoy_listeners "github.com/kumahq/kuma/pkg/xds/envoy/listeners"
+	"github.com/kumahq/kuma/pkg/xds/envoy/names"
+)
+
+// OriginInbound is a marker to indicate by which ProxyGenerator resources were generated.
+const OriginDNS = "dns"
+
+type DNSGenerator struct {
+}
+
+func (g DNSGenerator) Generate(ctx xds_context.Context, proxy *core_xds.Proxy) (*core_xds.ResourceSet, error) {
+	dnsPort := proxy.Metadata.GetDNSPort()
+	emptyDnsPort := proxy.Metadata.GetEmptyDNSPort()
+	if dnsPort == 0 || emptyDnsPort == 0 {
+		return nil, nil
+	}
+
+	if proxy.APIVersion == envoy_common.APIV2 {
+		return nil, nil // DNS is not available for API V2
+	}
+
+	if proxy.Dataplane.Spec.GetNetworking().GetTransparentProxying() == nil {
+		return nil, nil // DNS only makes sense when transparent proxy is used
+	}
+
+	vips := g.computeVIPs(ctx, proxy)
+	listener, err := envoy_listeners.NewListenerBuilder(proxy.APIVersion).
+		Configure(envoy_listeners.InboundListener(names.GetDNSListenerName(), proxy.Dataplane.GetIP(), dnsPort, core_xds.SocketAddressProtocolUDP)).
```

Comment:
Interesting, why it is an inbound listener? why not out bound?

Reply:
It is an extra server that accepts the traffic so I'd say it's rather inbound than outbound